packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,893 @@
+///
+/// ## Discussion
+///
+/// Due to Flutter's threading model and restrictions in the APIs of the
+/// platforms Flutter runs on, restoration data must be stored in the buckets
+/// proactively as described above. When the operating system asks for the
+/// restoration data, it will do so on the platform thread expecting a
+/// synchronous response. To avoid the risk of deadlocks, the platform thread
+/// cannot block and call into the UI thread (where the dart code is running) to
+/// retrieve the restoration data. For this reason, the [RestorationManager]
+/// always sends the latest copy of the restoration data from the UI thread over
+/// to the platform thread whenever it changes. That way, the restoration data
+/// is always ready to go on the platform thread when the operating system needs
+/// it.
+///
+/// See also:
+///
+///  * [ServicesBinding.restorationManager], which holds the singleton instance
+///    of the [RestorationManager] for the currently running application.
+///  * [RestorationBucket], which make up the restoration data hierarchy.
+///  * [RestorationId], which identifies a value or a child uniquely within a
+///    [RestorationBucket].
+///  * [RestorationMixin], which uses [RestorationBucket]s behind the scenes
+///    to make [State] objects of [StatefulWidget]s restorable.
+class RestorationManager {
+  /// The root of the [RestorationBucket] hierarchy containing the restoration
+  /// data.
+  ///
+  /// Child buckets can be claimed from this bucket via
+  /// [RestorationBucket.claimChild]. If the [RestorationManager] has been asked
+  /// to restore the application to a previous state, these buckets will contain
+  /// the previously stored data. Otherwise the root bucket (and all children
+  /// claimed from it) will be empty.
+  ///
+  /// Like any [RestorationBucket], the root bucket informs its listeners if it
+  /// has been replaced with a new root bucket. This happens when new
+  /// restoration data has been provided to the [RestorationManager] to restore
+  /// the application to a different state. In response to the notification,
+  /// listeners must stop using the old root bucket and obtain the new one via
+  /// this getter ([rootBucket] will have been updated to return the new bucket
+  /// just before the old root bucket informs its listeners).
+  ///
+  /// The restoration data describing the current bucket hierarchy is retrieved
+  /// asynchronously from the engine the first time the root bucket is accessed
+  /// via this getter. After the data has been copied over from the engine, this
+  /// getter will return a [SynchronousFuture], that immediately resolves to the
+  /// root [RestorationBucket].
+  ///
+  /// See also:
+  ///
+  ///  * [RootRestorationScope], which makes the root bucket available in the
+  ///    [Widget] tree.
+  Future<RestorationBucket> get rootBucket {
+    if (!_isListeningForEngineUpdates) {
+      SystemChannels.restoration.setMethodCallHandler(_methodHandler);
```

Comment:
Instead of doing this, use [`checkMethodCallHandler`](https://master-api.flutter.dev/flutter/services/MethodChannel/checkMethodCallHandler.html), or make `_isListeningForEngineUpdates` just make that call.

Reply:
Done.